File: 11_MinNumberInRotatedArray/Solution.py
# -*- coding:utf-8 -*-
# v-2 二分法-递归
class Solution:
    def minNumberInRotateArray(self, rotateArray):
        # write code here
        # 异常情况：数组为空或者为none
        if rotateArray is None or len(rotateArray)==0 :
            return 0
        # 特殊情况：数组没有旋转或者只有一个值
        p = 0
        q = len(rotateArray)-1
        if rotateArray[p] < rotateArray[q] or p==q:
            return rotateArray[p]
        # 正常情况
        return self.minNumberInRotate_equal(rotateArray[:])
        
    def minNumberInRotate(self, rotateArray):
        p = 0
        q = len(rotateArray)-1
        mid = (p+q)/2
        # 正常情况
        # 结束条件
        if q-p == 1:
            return rotateArray[q]
        if rotateArray[p] <= rotateArray[mid]:
            return self.minNumberInRotate(rotateArray[mid:(q+1)])
        else:
            return self.minNumberInRotate(rotateArray[p:(mid+1)])
    
    def minNumberInRotate_equal(self, rotateArray):
        p =0
        q = len(rotateArray)-1
        mid = (p+q)/2
        # 特殊情况：有相同的数字，无法判断mid属于哪里
        if rotateArray[p]==rotateArray[mid] and rotateArray[p]==rotateArray[q] :
            if rotateArray[p] < rotateArray[p+1] :
                return rotateArray[p]
            else:
                return self.minNumberInRotate_equal(rotateArray[1:])
        else:
            return self.minNumberInRotate(rotateArray[p:(q+1)])

File: 11_MinNumberInRotatedArray/Solution.java
# v-2 二分法
import java.util.ArrayList;
public class Solution {
    public int minNumberInRotateArray(int [] array) {
        if (array==null || array.length==0){
            return 0;
        }
        int p = 0;
        int q = array.length-1;
        if (array[p] < array[q]){
            return array[p];
        }
        
        if (array[p]==array[q]){
            return minNumberArray_equal(array, p, q);
        }
        return minNumberArray(array, p, q);
    }
    
    public int minNumberArray(int [] array, int p, int q){
        // 正常情况，前后两个递增数组
        // 结束条件
        if (q-p==1){
            return array[q];
        }
        
        int mid = (p+q)/2;
        if (array[p]<=array[mid]){
            return minNumberArray(array, mid, q);
        }
        else{
            return minNumberArray(array, p, mid);
        }
    }
    
    public int minNumberArray_equal(int [] array, int p, int q){
        // 特殊情况，array[p]==array[q]
        // 结束条件
        if (array[p]==array[q]){
            if (array[p+1]<array[p]){
                return array[p+1];
            }
            else{
                return minNumberArray_equal(array, p+1, q);
            }
        }
        else{
            return minNumberArray(array, p, q);
        }
    }
}


# v-2 二分法 优化了一些代码和边界条件

import java.util.ArrayList;
public class Solution {
    public int minNumberInRotateArray(int [] array) {
        if (array==null || array.length==0){
            return 0;
        }
        int p = 0;
        int q = array.length-1;
        // 特殊情况，没有发生旋转或者只有一个元素
        if (array[p] < array[q]||p==q){
            return array[p];
        }
        
        return minNumberArray_equal(array, p, q);
    }
    
    public int minNumberArray(int [] array, int p, int q){
        // 正常情况，前后两个递增数组
        // 结束条件
        if (q-p==1){
            return array[q];
        }
        
        int mid = (p+q)/2;
        if (array[p]<=array[mid]){
            return minNumberArray(array, mid, q);
        }
        else{
            return minNumberArray(array, p, mid);
        }
    }
    
    public int minNumberArray_equal(int [] array, int p, int q){
        // 特殊情况，array[p]==array[q]
        // 结束条件
        int mid = (p+q)/2;
        if (array[p]==array[q] && array[p]==array[mid]){
            if (array[p+1]<array[p]){
                return array[p+1];
            }
            else{
                return minNumberArray_equal(array, p+1, q);
            }
        }
        else{
            return minNumberArray(array, p, q);
        }
    }
}
